File: src/lib/shippingFee.ts
/**
 * 배송비 계산 (화면용).
 *
 * 서버 shared/shipping_fee.py 와 같은 규칙이다. 주문 전에 손님에게 보여줄
 * 금액을 미리 계산하려면 화면에도 같은 계산이 있어야 한다. 실제 청구는
 * 서버가 다시 계산하므로 여기 값이 틀려도 금액이 잘못 나가지는 않는다.
 */
export interface ShippingTier {
  qty: number
  fee: number
}

export function normalizeTiers(raw: unknown): ShippingTier[] {
  if (!Array.isArray(raw)) return []
  return raw
    .map((item) => {
      if (!item || typeof item !== 'object') return null
      const qty = Number((item as any).qty)
      const fee = Number((item as any).fee)
      if (!Number.isInteger(qty) || qty <= 0 || !Number.isFinite(fee) || fee < 0) return null
      return { qty, fee: Math.round(fee) }
    })
    .filter((t): t is ShippingTier => t !== null)
    .sort((a, b) => a.qty - b.qty)
}

/**
 * 이 수량의 배송비. qty 는 '이 수량까지' 를 뜻한다.
 *
 * 표에 없는 큰 수량은 마지막 구간을 되풀이한다. 3box 까지만 정해 뒀는데
 * 7box 를 시키면 3box 묶음 세 번으로 본다.
 * 표가 비어 있으면 0 — 상품가에 포함된 것으로 본다.
 */
export function shippingFeeFor(raw: unknown, quantity: number): number {
  const tiers = normalizeTiers(raw)
  if (tiers.length === 0 || quantity <= 0) return 0
  for (const tier of tiers) if (quantity <= tier.qty) return tier.fee
  const last = tiers[tiers.length - 1]
  return last.fee * Math.ceil(quantity / last.qty)
}

File: server-py/app/shared/shipping_fee.py
"""
배송비 계산.

상품마다 수량 구간별 요금표를 둔다.
    [{"qty": 1, "fee": 5000}, {"qty": 2, "fee": 7000}, {"qty": 3, "fee": 8000}]
qty 는 '이 수량까지' 를 뜻한다. 2박스를 시키면 7,000원이지 5,000×2 가 아니다.

표가 비어 있으면 0 이다 — 지금까지처럼 상품가에 포함된 것으로 본다.
기존 상품이 그대로 동작해야 하기 때문이다.
"""
from typing import Any


def normalize_tiers(raw: Any) -> list[tuple[int, int]]:
    """[{qty, fee}] 를 (수량, 요금) 목록으로. 잘못된 항목은 버린다."""
    if not isinstance(raw, list):
        return []
    tiers: list[tuple[int, int]] = []
    for item in raw:
        if not isinstance(item, dict):
            continue
        try:
            qty, fee = int(item["qty"]), int(item["fee"])
        except (KeyError, TypeError, ValueError):
            continue
        if qty > 0 and fee >= 0:
            tiers.append((qty, fee))
    return sorted(tiers)


def fee_for(raw: Any, quantity: int) -> int:
    """
    이 수량의 배송비.

    표에 없는 큰 수량은 마지막 구간을 되풀이한다. 3box 까지만 정해 뒀는데
    7box 를 시키면 3box 묶음 세 번으로 본다. 실제로 상자를 나눠 보내야
    하므로 요금도 그만큼 든다.
    """
    tiers = normalize_tiers(raw)
    if not tiers or quantity <= 0:
        return 0
    for qty, fee in tiers:
        if quantity <= qty:
            return fee
    max_qty, max_fee = tiers[-1]
    groups = -(-quantity // max_qty)   # 올림
    return max_fee * groups
